### PR TITLE
Transition to purescript-cardano-serialization-lib

### LINKED ISSUE #1534: Switch to purescript-cardano-serialization-lib
https://github.com/mlabs-haskell/purescript-cardano-serialization-lib

### PR DESCRIPTION
**Pull request by klntsky**
Closes #1534

### Pre-review checklist

- [ ] All code has been formatted using our config (`make format`)
- [ ] Any new API features or modification of existing behavior are covered with tests
- [ ] The template (`templates/ctl-scaffold`) has been [updated](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/development.md#maintaining-the-template)
- [ ] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Changed`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: src/Internal/BalanceTx/Collateral/Select.purs [hunk 1]
```diff
@@ -158,16 +159,27 @@ asTxUnspentOutput
...
+
+unsafeFromMaybe :: forall a. String -> Maybe a -> a
+unsafeFromMaybe e a = case a of
+  Nothing ->
+    unsafePerformEffect $ throw $ "unsafeFromMaybe: impossible happened: "
+      <> e
+      <> " (please report as bug at "
+      <> bugTrackerLink
+      <> " )"
+  Just v -> v
```
Comment (errfrom) [inline src/Internal/BalanceTx/Collateral/Select.purs:178]: I think we should extract this function to Utils / Helpers module and use it in place of uninformative `fromJust` across the entire codebase

**Review by errfrom — COMMENTED** [review]
Partial review, to be continued.
Tremendous work @klntsky!

File: src/Internal/Types/ScriptLookups.purs [hunk 2]
```diff
@@ -46,17 +49,15 @@ import Data.Show.Generic (genericShow)
...
-  , txOutputs ::
-      Map TransactionInput Plutus.TransactionOutputWithRefScript -- Unspent outputs that the script may want to spend
-  , scripts ::
-      Array Validator -- Script validators
-  , datums :: Map DataHash Datum --  Datums that we might need
+  { plutusMintingPolicies :: Array PlutusScript
+  , nativeMintingPolicies :: Array NativeScript
+  , txOutputs :: UtxoMap
+  , scripts :: Array PlutusScript -- Script validators
+  , datums :: Map DataHash PlutusData --  Datums that we might need
```
Comment (errfrom) [inline src/Internal/Types/ScriptLookups.purs:56]: "Datums that we might need" - such comments are not particularly useful

File: src/Internal/Types/TxConstraints.purs [hunk 3]
```diff
@@ -472,79 +452,72 @@ mustPayToNativeScriptAddress nsHash credential vl =
...
 -- | Create the given amount of the currency.
 -- | The amount to mint must not be zero.
 mustMintCurrency
-  :: forall (i :: Type) (o :: Type)
-   . MintingPolicyHash
-  -> TokenName
-  -> BigInt
+  :: ScriptHash
+  -> AssetName
+  -> Int.Int
```
Comment (errfrom) [inline src/Internal/Types/TxConstraints.purs:478]: `Int.int` can have a `Ring` instance

File: src/Internal/Types/TxConstraints.purs [hunk 4]
```diff
@@ -696,75 +666,3 @@ mustSatisfyAnyOf =
...
-  :: TxConstraints -> Value
-mustSpendAtLeastTotal =
-  foldr (join <<< f) mempty <<< _.constraints <<< unwrap
-  where
-  f :: TxConstraint -> Value
-  f (MustSpendAtLeast v) = v
-  f _ = mempty
-
--- | The minimum `Value` that satisfies all `MustProduceAtLeast` constraints
-mustProduceAtLeastTotal
```
Comment (errfrom) [inline src/Internal/Types/TxConstraints.purs:737]: `mustProduceAtLeastTotal` and `mustSpendAtLeastTotal` are still mentioned in doc/test-plan.md

File: src/Internal/Lens.purs [hunk 5]
```diff
@@ -0,0 +1,153 @@
+module Ctl.Internal.Lens
```
[the remark this answers is not in the document]
Comment (klntsky) [reply src/Internal/Lens.purs:1]: The dependency footprint of lenses is huge (and runtime cost)

File: src/Internal/Cardano/Types/Transaction.purs [hunk 6]
```diff
@@ -1,1007 +0,0 @@
-module Ctl.Internal.Cardano.Types.Transaction
```
[the remark this answers is not in the document]
Comment (klntsky) [reply src/Internal/Cardano/Types/Transaction.purs:1]: It's not possible. The type class has to be two-parameter (to connect two types), and PS requires the instances to be declared either in the module with the type(s) or in the module with the class.

File: src/Internal/BalanceTx/CoinSelection.purs [hunk 7]
```diff
@@ -229,18 +221,18 @@ selectedBalance = balance <<< _.selectedUtxos <<< unwrap
...
 -- |
 -- | Taken from cardano-wallet:
 -- |https://github.com/input-output-hk/cardano-wallet/blob/a61d37f2557b8cb5c47b57da79375afad698eed4/lib/wallet/src/Cardano/Wallet/CoinSelection/Internal/Balance.hs#L2175
 selectedCoinQuantity :: SelectionState -> BigInt
-selectedCoinQuantity = Value.valueToCoin' <<< selectedBalance
+selectedCoinQuantity = Val.getCoin <<< selectedBalance
 
 -- | Returns the output references of the selected utxos.
 selectedInputs :: SelectionState -> Set TransactionInput
-selectedInputs = Set.fromFoldable <<< Map.keys <<< view _selectedUtxos
```
[the remark this answers is not in the document]
Comment (klntsky) [reply src/Internal/BalanceTx/CoinSelection.purs:243]: ?

File: src/Internal/Lens.purs (hunk 5 above)
[the remark this answers is not in the document]
Comment (klntsky) [reply src/Internal/Lens.purs:1]: I would remove them from CTL too honestly if I had time.

File: src/Internal/BalanceTx/FakeOutput.purs [hunk 8]
```diff
@@ -19,14 +18,14 @@ fakeOutputWithValue amount =
...
+    , datum: Nothing
     , scriptRef: Nothing
     }
 
-fakeOutputWithNonAdaAssets :: NonAdaAsset -> TransactionOutput
-fakeOutputWithNonAdaAssets =
-  fakeOutputWithValue <<< mkValue mempty
+fakeOutputWithMultiAssets :: MultiAsset -> TransactionOutput
+fakeOutputWithMultiAssets =
+  fakeOutputWithValue <<< mkValue Coin.zero
```
[the remark this answers is not in the document]
Comment (klntsky) [reply src/Internal/BalanceTx/FakeOutput.purs:31]: Sadly, it requires `Partial`. I think having to use `unsafePartial` for `zero` is a worse API than explicit named `zero` value.

**Review by errfrom — APPROVED** [review]
LGTM! Can't wait to upgrade the projects to use this revised version of CTL, it's much cleaner. My only concern is the removal of the Cardano <> Plutus conversion functionality, which could be bad for UX.

File: src/Internal/Lens.purs (hunk 5 above)
[the remark this answers is not in the document]
Comment (errfrom) [reply src/Internal/Lens.purs:1]: Makes sense, should we open an issue for that?

File: src/Internal/BalanceTx/CoinSelection.purs (hunk 7 above)
[the remark this answers is not in the document]
Comment (errfrom) [reply src/Internal/BalanceTx/CoinSelection.purs:243]: Nvm, it's just me doing the conversion from Set to Set

File: CHANGELOG.md [hunk 9]
```diff
@@ -64,16 +64,48 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
...
+- this package
+- `purescript-cardano-types` - (`Cardano.Types.*`) - domain types for Cardano ledger
+- `purescript-cardano-serialization-lib` - (`Cardano.Serialization.Lib.*`) - PureScript wrapper for CSL
+- `purescript-plutus-types` - (`Cardano.Plutus.*`) - Plutus-style domain types used only for `ToData`/`FromData` instances.
+
+See the [cardano-purescript repo](https://github.com/klntsky/cardano-purescript) for a complete list of new packages.
+
+These packages maintain their own public API, so the interfaces you should use are not necessarily namespaced under `Contract.*` anymore. E.g. `Cardano.Types.*` from `purescript-cardano-types` is public.
+
+Starting from this version, CTL does not use Plutus-domain types anymore. There are no more Plutus domain <-> Cardano domain conversion functions. This change comes with some downsides (e.g. the need to provide NetworkId in `Address`es), but the benefit of not caring about conversions for basic functionality will hopefully be more noticeable.
```
[the remark this answers is not in the document]
Comment (klntsky) [reply CHANGELOG.md:78]: Agreed. Will restore it now.

**Pull request merged**